code/game/objects/structures/doors.dm

```diff
@@ -0,0 +1,902 @@
+			return FALSE
+
+		var/obj/item/tool/weldingtool/welder = I
+		if(!welder.check_fuel() || !welder.isOn())
+			return FALSE
+
+		if(tool_action_time && !do_after(user, tool_action_time, NONE, src, BUSY_ICON_BUILD))
+			return FALSE
+
+		if(!welder.remove_fuel(M = user))
+			return FALSE
+
+		do_deconstruction()
+		return TRUE
+
+	//Man, these welder procs should really be consolidated
+	var/obj/item/tool/weldingtool/welder = I
+	if(!welder.check_fuel() || !welder.isOn())	//Make sure the welder is on and has fuel
+		return FALSE
+
+	if(tool_action_time && !do_after(user, tool_action_time, NONE, src, BUSY_ICON_BUILD))
+		return FALSE
+
+	if(!welder.remove_fuel(M = user))	//Make sure the welder still has fuel after it's done
+		return FALSE
+
+	do_deconstruction()
+	return TRUE
+
+/obj/structure/door/weld_cut_act(mob/living/user, obj/item/I)
+	if(!length(construction_steps))
+		return FALSE
+
+	if(!CHECK_BITFIELD(door_flags, DOOR_FORCEABLE))
+		to_chat(user, span_warning("[src] is too sturdy to disassemble!"))
+		return FALSE
+
+	if(lock?.locked && CHECK_BITFIELD(door_flags, DOOR_OPEN))
+		balloon_alert(user, "Unlock it first!")
+		return FALSE
+
+	if(I.tool_behaviour != construction_steps[construction_steps_index])
+		if(construction_steps_index == 1 || (I.tool_behaviour != construction_steps[construction_steps_index - 1]))
+			return FALSE
+
+		var/obj/item/tool/pickaxe/plasmacutter/cutter = I
+		if(!(cutter.cell.charge >= cutter.cell.charge_amount) || !cutter.powered)
+			cutter.fizzle_message(user)
+			return FALSE
+
+		playsound(src, cutter.cutting_sound, 25, 1)
+		cutter.eyecheck(user)
+		if(tool_action_time && !do_after(user, tool_action_time, NONE, src, BUSY_ICON_BUILD))
+			return FALSE
+
+		if(!(cutter.cell.charge >= cutter.cell.charge_amount) || !cutter.powered)
+			cutter.fizzle_message(user)
+			return FALSE
+
+		cutter.use_charge(user)
+		do_construction()
+		return TRUE
+
+	//This is even worse than the welder procs... at least it had procs
+	var/obj/item/tool/pickaxe/plasmacutter/cutter = I
+	if(!(cutter.cell.charge >= cutter.cell.charge_amount) || !cutter.powered)
+		cutter.fizzle_message(user)
+		return FALSE
+
+	playsound(src, cutter.cutting_sound, 25, 1)
+	cutter.eyecheck(user)
+	if(tool_action_time && !do_after(user, tool_action_time, NONE, src, BUSY_ICON_BUILD))
+		return FALSE
+
+	if(!(cutter.cell.charge >= cutter.cell.charge_amount) || !cutter.powered)
+		cutter.fizzle_message(user)
+		return FALSE
+
+	cutter.use_charge(user)
+	do_deconstruction()
+	return TRUE
+
+/obj/structure/door/multitool_act(mob/living/user, obj/item/I)
+	return handle_tools(user, I)
+
+/obj/structure/door/analyzer_act(mob/living/user, obj/item/I)
+	return handle_tools(user, I)
+
+///Reduce line copy and pasting by having one proc do (most of) them
+/obj/structure/door/proc/handle_tools(mob/living/user, obj/item/tool)
+	if(!length(construction_steps))	//No message needed because the user would know if they examined the door!
+		return FALSE
+
```

Comment:
Welder being snowflaked out is :/
The attack chain seperating out behaviour just for wrappers that feed into this proc is :/, would look again how mech does it

Reply:
Fair point. I was going to say I could just move it to `tool_attack_chain` but that is on the item not the targeted object. What do you propose? Because `handle_tools` is meant to be agnostic of the tool type, just needs to be the required tool to proceed. Only reason it's a proc is because repeated code icky.

As for welders, yeah `welder_repair_act` should have been used there. But in my opinion is still that the bulk of the code should be a function on the welder, not the target. 🤷 Same for plasma cutters.

Reply:
Nevermind I remembered why I needed this. Needed to make sure welding tool was on, had fuel, and then remove that fuel.